man/bppancestor.1
.\" SPDX-FileCopyrightText: The Bio++ Development Group
.\"
.\" SPDX-License-Identifier: CECILL-2.1


.TH BPPANCESTOR 1 LOCAL

.SH NAME

bppml - Ancestral sequence reconstruction with Bio++

.SH SYNOPSIS

.B bppancestor [options]

.SH AVAILABILITY

All UNIX flavors

.SH DESCRIPTION

bppancestor performs ancestral sequence reconstructions under a large variety of models, including non-homogeneous ones.

.SH OPTIONS

You should refer to 'info bppsuite' or to the online manual of bppsuite for a complete list of available options.

.TP 5

--noninteractive

generates output for redirection in a file.

.TP

param=file

reads a file for loading options

.SH AUTHOR

The Bio++ Development Team.


ChangeLog
19/03/18 -*- Version 2.4.0 -*-

19/02/18 Julien Dutheil
* Bug fixed in bppPopStats: dN and dS were inverted!

10/12/17 -*- Version 2.3.2 -*-

06/11/17 Julien Dutheil
* Added estimation of kappa to bppPopStats + bugs fixed

06/06/17 -*- Version 2.3.1 -*- 

10/05/17 -*- Version 2.3.0 -*- 

21/12/16 Julien Dutheil
* BppPhySamp migrated to its own package.

17/12/16 Julien Dutheil
* Improved bppPopStats, outgroup analyses, MK table and output to file.

06/10/16 Julien Dutheil
* bppSeqGen now takes Hudson's MS ARG files as input.
* new input.tree.scale option to adjust branch lengths.

22/05/16 Julien Dutheil
* bpppopstats now computes codon statistics per site (e.g. PiN, PiS)
* Added example for bpppopstats.

29/06/15 Julien Dutheil
* New program bppPopStats

28/01/15 Julien Dutheil
* PhySamp now outputs the correct tree!

25/11/14 Julien Dutheil
* bppalnscore output contiguous segments in mase selection.
* bppalnscore site selection works with codon alphabets.

24/11/14 Julien Dutheil
* bppSeqMan can now read alignments and perform site selection priori to any analysis.

28/09/14 -*- Version 2.2.0 -*- 

16/09/14 Julien Dutheil
* bppSeqGen supports generic characters as ancestral states in info file (bug #87).

12/08/14 Julien Dutheil
* bppSeqGen supports simulation under Markov-Modulated models.

06/06/14 Julien Dutheil
* bppPhySamp now output trees as well.

17/04/14 Julien Dutheil
* bppSeqMan: Translate uses global genetic_code argument instead of local 'code' argument.
* All programs use the --seed argument to set random seed for reproduceable results.

08/03/13 -*- Version 0.8.0 -*-

22/01/13 Julien Dutheil
* Compiles with -Wconversion

14/01/13 Mathieu Groussin
* Added COaLA model for proteins in bppML and bppSeqGen.

23/11/12 Laurent Guéguen
* New bppMixedLikelihood program

08/02/12 -*- Version 0.7.0 -*-

21/12/11 Julien Dutheil
* New bppAlnScore program
* Improved bppSeqMan: protein translation now works + more stop codons removal options.

17/06/11 Julien Dutheil
* Improved bppancestor, also allows to compute frequencies for leaves.

11/06/11 -*- Version 0.6.2 -*-
* Small bug fixed + new packages.

03/06/10 -*- Version 0.6.0 -*-
* Compatibility with Bio++ 2.0.0.

03/06/10 Julien Dutheil
* New method in bppPhySamp + new, faster, implementation.

25/03/10 -*- Version 0.5.0 -*-

22/08/09 Julien Dutheil
* Added CMake configuration files, with twick for building the manual (assumes
		texinfo is installed).

22/07/09 Julien Dutheil
* Added program bppTreeDraw.

30/06/09 Julien Dutheil
* BppAncestor: now output expected ancestral states frequencies.
* BppAncestor: new method to compute EB estimates of ancestral states frequencies.

10/06/09 -*- Version 0.4.0 -*-

22/05/09 Julien Dutheil
* Keyval syntax for sequence format description.
* Documentation and example update.
* SeqMan now uses the SequenceApplicationTools method.

15/05/09 Julien Dutheil
* Compatibility update with the libs, new keyval syntax for model
specification.
* Documentation up to date.

02/02/09 Julien Dutheil
* bppConsense now outputs the number of leaves of the input tree.
* New progress bar in bppConsense.

08/01/09 Julien Dutheil
* Added texinfo documentation.
* More bootstrap options in BppDist (the same as in BppML).

25/09/08 -*- Version 0.3.1 -*-

07/12/08 Bastien Boussau & Julien Dutheil
* Bug removed in non-homogeneous models with covarions.

09/10/08 Julien Dutheil
* New action 'invert' in bppSeqMan.

25/09/08 -*- Version 0.3.0 -*-

20/09/08 Julien Dutheil
* New program bppAncestor to reconstruct acestral states and rates.
* New +F models from PhylLib. 

02/07/08 Laurent Guegen
* Bug fixed in bppML: frequencies can be computed from the data with
non-homogeneous models.

19/06/08 Julien Dutheil
* Shell variables are now 'global' in option files.

24/04/08 -*- Version 0.2.3 -*-
* This version requires the CVS version of Bio++ to be built.

03/03/08 Julien Dutheil
* Bug fixed in keep_complete method of BppSeqMan.

27/03/08 Julien Dutheil
* Bug fixed when printing parameters of non-homogeneous models in bppML.

22/01/08 -*- Version 0.2.2 -*-

28/02/08 Celine Scronavacca
* Bug fixed in ReRoot.

22/01/08 -*- Version 0.2.1 -*-

06/02/08 Celine Scornavacca
* Bug fixed in ReRoot + correction of version number in bppSeqGen.

22/01/08 -*- Version 0.2.0 -*-

13/01/08 Julien Dutheil
* More options in bppml: possibility to chose between simple and double
recursion and data compression method.
* Simple recursion method improved.

11/01/08 Julien Dutheil
* Compatibility update
* bppML now output more handy parameter values.

31/12/07 Julien Dutheil
* New program Phylo Sampler.

07/11/07 Julien Dutheil
* Compatibility update with Bio++ cvs.
* Bug fixed in BppSeqGen.
* BppSeqGen can now simulate data from a set of trees (ancestral
  recombination graph).

25/10/07 Julien Dutheil
* Several bug fixed in BppML and BppSeqGen with non-homogeneous models.
* Examples updated + new example option files for non-homogeneous models.

05/10/07 Bastien Boussau & Julien Dutheil
* Non-homogenous sequence simulation in BppSeqGen.
* Model options in files have changed to suit the non-homogeneous case.
  The covarion option now is fully part of the model options, following the
	new syntax 'model=modelname+covarionname', eg 'T92+G2001'.

03/10/07 Julien Dutheil
* New program Bio++ Sequence Manipulator

21/09/07 Julien Dutheil
* Bug fixed in BppDist: branch length are indeed not estimated when parameters
  to ignore contains "BrLen".
* Stop condition when using iterations is now based on topological distance
  instead of likelihood comparisons (phyllib).
* New option optimization.topology_nni.method = [fast|better|phyml]

17/06/07 -*- Version 0.1.0 -*-

16/07/06 Julien Dutheil
* New option bootstrap.verbose for bppML.
* Bug fixed in bootstrap computation.

06/06/06 Julien Dutheil
* New program bppConsense.

05/05/07 Julien Dutheil
* First alpha release of bppsuite.

-*- Old bppML ChangeLog -*-

23/04/07 Julien Dutheil

bppSuite/bppAncestor.cpp
// SPDX-FileCopyrightText: The Bio++ Development Group
//
// SPDX-License-Identifier: CECILL-2.1

// From the STL:
#include <iostream>
#include <iomanip>

using namespace std;

// From bpp-core:
#include <Bpp/Version.h>
#include <Bpp/Numeric/DataTable.h>

// From bpp-seq:
#include <Bpp/Seq/SiteTools.h>
#include <Bpp/Seq/Container/SequenceContainerTools.h>
#include <Bpp/Seq/Io/BppOAlignmentWriterFormat.h>

// From bpp-phyl:
#include <Bpp/Phyl/App/BppPhylogeneticsApplication.h>
#include <Bpp/Phyl/App/PhylogeneticsApplicationTools.h>
#include <Bpp/Phyl/Likelihood/MarginalAncestralReconstruction.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/OneProcessSequencePhyloLikelihood.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/AlignedPhyloLikelihoodSet.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/SingleProcessPhyloLikelihood.h>

using namespace bpp;

/******************************************************************************/

int main(int args, char** argv)
{
  cout << "******************************************************************" << endl;
  cout << "*     Bio++ Ancestral Sequence Reconstruction, version " << BPP_VERSION << "     *" << endl;
  cout << "* Authors: J. Dutheil                       Created on: 10/09/08 *" << endl;
  cout << "*          B. Boussau                       Last Modif: 25/09/14 *" << endl;
  cout << "*          L. Guéguen                       Last Modif: 22/12/14 *" << endl;
  cout << "******************************************************************" << endl;
  cout << endl;

  try
  {
    BppPhylogeneticsApplication bppancestor(args, argv, "bppancestor");

    if (args == 1)
    {
      bppancestor.help("bppAncestor");
      return 0;
    }

    bppancestor.startTimer();

    Context context;
    map<string, string> allParams = bppancestor.getParams();
    map<string, string> unparsedParams;

    shared_ptr<const Alphabet> alphabet(bppancestor.getAlphabet());
    shared_ptr<const GeneticCode> gCode(bppancestor.getGeneticCode(alphabet));

    // Missing check
    //  if (model->getName() != "RE08") SiteContainerTools::changeGapsToUnknownCharacters(*sites);

    // get the result phylo likelihood
    auto mSitesuniq = bppancestor.getConstAlignmentsMap(alphabet, true);

    const std::map<size_t, std::shared_ptr<const AlignmentDataInterface >> mSites = PhylogeneticsApplicationTools::uniqueToSharedMap<const TemplateAlignmentDataInterface<string>>(mSitesuniq);

    auto mpTree = bppancestor.getPhyloTreesMap(mSites, unparsedParams);
    shared_ptr<SubstitutionProcessCollection> SPC = bppancestor.getCollection(alphabet, gCode, mSites, mpTree, unparsedParams);
    auto mSeqEvoltmp = bppancestor.getProcesses(SPC, unparsedParams);
    auto mSeqEvol = PhylogeneticsApplicationTools::uniqueToSharedMap<SequenceEvolution>(mSeqEvoltmp);

    auto mPhyl = bppancestor.getPhyloLikelihoods(context, mSeqEvol, SPC, mSites);

    // retrieve Phylo 0, aka result phylolikelihood

    if (!mPhyl->hasPhyloLikelihood(0))
      throw Exception("Missing phyloLikelihoods.");

    auto tl = (*mPhyl)[0];

    bppancestor.fixLikelihood(alphabet, gCode, tl);

    bppancestor.displayParameters(*tl, false);

    ApplicationTools::displayMessage("");

    //////////////////////////////////////
    // Reconstruct ancestral sequences:


    /// map of the Single Data Process
    map<size_t, shared_ptr<AbstractSingleDataPhyloLikelihood>> mSD;

    if (dynamic_pointer_cast<AbstractSingleDataPhyloLikelihood>(tl) != NULL)
      mSD[1] = dynamic_pointer_cast<AbstractSingleDataPhyloLikelihood>(tl);
    else
    {
      auto sOAP = dynamic_pointer_cast<AlignedPhyloLikelihoodSetInterface>(tl);
      if (sOAP)
      {
        const vector<size_t>& nSD = sOAP->getNumbersOfPhyloLikelihoods();

        for (size_t iSD = 0; iSD < nSD.size(); iSD++)
        {
          auto pASDP = dynamic_pointer_cast<AbstractSingleDataPhyloLikelihood>(sOAP->getPhyloLikelihood(nSD[iSD]));
          if (pASDP != NULL)
            mSD[nSD[iSD]] = pASDP;
        }
      }
    }

    ////////////////////////
    /// Options

    // Ancestral information
    // Sites

    string outputSitesFile = ApplicationTools::getAFilePath("output.sites.file", allParams, false, false);

    bool probs(false);

    if (outputSitesFile != "none")
    {
      probs = ApplicationTools::getBooleanParameter("output.sites.probabilities", allParams, false, "", true, 1);
      if (!probs)
        probs = ApplicationTools::getBooleanParameter("asr.probabilities", allParams, false, "", true, 1);

      ApplicationTools::displayResult("Output site probabilities", probs ? "yes" : "no");
    }

    // Nodes

    bool addNodesExtant = false;
    string outputNodesFile = ApplicationTools::getAFilePath("output.nodes.file", allParams, false, false, "", false, "none", 1);

    if (outputNodesFile != "none")
    {
      addNodesExtant = ApplicationTools::getBooleanParameter("output.nodes.add_extant", allParams, false, "", true, 1);
      ApplicationTools::displayResult("Output extant nodes", addNodesExtant ? "yes" : "no");
    }

    // ASR

    string sequenceFilePath = ApplicationTools::getAFilePath("asr.sequence.file", allParams, false, false, "", false, "none", 1);

    if (sequenceFilePath == "none")
      sequenceFilePath = ApplicationTools::getAFilePath("output.sequence.file", allParams, false, false, "", false, "none", 1);

    string sequenceFormat = "";
    bool sample = false;
    unsigned int nbSamples = 0;
    bool addSitesExtant = false;

    if (sequenceFilePath != "none")
    {
      sequenceFormat   = ApplicationTools::getStringParameter("asr.sequence.format", allParams, "none", "", false, 1);
      if (sequenceFormat == "none")
        sequenceFormat   = ApplicationTools::getStringParameter("output.sequence.format", allParams, "Fasta", "", false, 1);

      sample = ApplicationTools::getBooleanParameter("asr.sample", allParams, false, "", true, 1);

      ApplicationTools::displayResult("Sample from posterior distribution", sample ? "yes" : "no");

      if (sample)
        nbSamples = ApplicationTools::getParameter<unsigned int>("asr.sample.number", allParams, 1, "", true, false);

      addSitesExtant = ApplicationTools::getBooleanParameter("asr.add_extant", allParams, false, "", true, 1);
      ApplicationTools::displayResult("ASR extant", addSitesExtant ? "yes" : "no");
    }


    /////////////////////////////////
    /// per Single Data Process

    for (auto& itm:mSD)
    {
      auto sPP = dynamic_pointer_cast<SingleProcessPhyloLikelihood>(itm.second);
      auto oPSP = dynamic_pointer_cast<OneProcessSequencePhyloLikelihood>(itm.second);

      if ((sPP == NULL) && (oPSP == NULL))
      {
        ApplicationTools::displayWarning("Multi Process ancestral reconstruction not implemented");
        ApplicationTools::displayWarning("for phyloLikelihood " + TextTools::toString(itm.first));
        continue;
      }

      auto pDR = sPP ? sPP->getLikelihoodCalculationSingleProcess() : oPSP->getLikelihoodCalculationSingleProcess();
      auto sites = std::shared_ptr<const AlignmentDataInterface>(sPP ? sPP->getData() : oPSP->getData());

      // Only Marginal reconstruction method

      VectorSequenceContainer vSC(alphabet);

      if (addSitesExtant)
      {
        auto vSC0 = dynamic_pointer_cast<const SiteContainerInterface>(sites);
        if (!vSC0)
          ApplicationTools::displayWarning("Output extant sequences not possible with probabilistic sequences.");
        // Keep only leaves
        const auto tree = sPP ? sPP->tree() : oPSP->tree();
        const auto leaves = tree->getAllLeavesNames();

        SequenceContainerTools::getSelectedSequences(*vSC0, leaves, vSC);
      }

      AncestralStateReconstruction* asr = new MarginalAncestralReconstruction(pDR);

      size_t nbStates = sPP ? sPP->getNumberOfStates() : oPSP->getNumberOfStates();

      const auto& stateMap = pDR->stateMap();

      ApplicationTools::displayMessage("\nPhylo " + TextTools::toString(itm.first));

      /////////////////////////////////////
      // Write sites infos to file:

      if (outputSitesFile != "none")
      {
        string outF = outputSitesFile + "_" + TextTools::toString(itm.first);
        ApplicationTools::displayResult(" Output file for sites", outF);
        ofstream out(outF.c_str(), ios::out);
        auto ttree(sPP ? sPP->tree() : oPSP->tree());
        vector<shared_ptr<PhyloNode>> nodes = ttree->getAllNodes();
        size_t nbNodes = nodes.size();

        // Get the class with maximum posterior probability:
        vector<size_t> classes = sPP ? sPP->getClassWithMaxPostProbPerSite() : oPSP->getClassWithMaxPostProbPerSite();
        // Get the posterior rate, i.e. rate averaged over all posterior probabilities:

        Vdouble rates = sPP ? sPP->getPosteriorRatePerSite() : oPSP->getPosteriorRatePerSite();

        // Get the ancestral sequences:
        vector<unique_ptr<Sequence>> sequences(nbNodes);
        vector<VVdouble*> probabilities(nbNodes);

        vector<string> colNames;
        colNames.push_back("Sites");
        colNames.push_back("is.complete");
        colNames.push_back("is.constant");
        colNames.push_back("lnL");
        colNames.push_back("rc");
        colNames.push_back("pr");

        for (size_t i = 0; i < nbNodes; i++)
        {
          shared_ptr<PhyloNode> node = nodes[i];
          auto nodeindex = ttree->getNodeIndex(node);
          colNames.push_back("max." + TextTools::toString(nodeindex));
          if (probs)
          {
            probabilities[i] = new VVdouble();

            // The cast will have to be updated when more probabilistic method will be available:
            sequences[i] = dynamic_cast<MarginalAncestralReconstruction*>(asr)->getAncestralSequenceForNode(nodeindex, probabilities[i], false);

            for (unsigned int j = 0; j < nbStates; j++)
            {
              colNames.push_back("prob." + TextTools::toString(nodeindex) + "." + stateMap.getStateDescription(j));
            }
          }
          else
            sequences[i] = asr->getAncestralSequenceForNode(nodeindex);
        }

        // Now fill the table:
        vector<string> row(colNames.size());
        auto infos = make_unique<DataTable>(colNames);

        for (size_t i = 0; i < sites->getNumberOfSites(); i++)
        {
          double lnL = sPP ? sPP->getLogLikelihoodForASite(i) : oPSP->getLogLikelihoodForASite(i);

          auto& currentSite = sites->site(i);
          int currentSitePosition = currentSite.getCoordinate();
          string isCompl = "NA";
          string isConst = "NA";
          try
          {
            isCompl = (SiteTools::isComplete(currentSite) ? "1" : "0");
          }
          catch (EmptySiteException& ex)
          {}
          try
          {
            isConst = (SiteTools::isConstant(currentSite) ? "1" : "0");
          }
          catch (EmptySiteException& ex)
          {}
          row[0] = (string("[" + TextTools::toString(currentSitePosition) + "]"));
          row[1] = isCompl;
          row[2] = isConst;
          row[3] = TextTools::toString(lnL);
          row[4] = TextTools::toString(classes[i]);
          row[5] = TextTools::toString(rates[i]);

          unsigned int k = 6;
          for (unsigned int j = 0; j < nbNodes; j++)
          {
            row[k] = sequences[j]->getChar(i);
            k++;
            if (probs)
            {
              for (unsigned int l = 0; l < nbStates; l++)
              {
                row[k] = TextTools::toString((*probabilities[j])[i][l]);
                k++;
              }
            }
          }

          infos->addRow(row);
        }

        DataTable::write(*infos, out, "\t");
      }


      /////////////////////////////////////
      // Write nodes infos to file:

      if (outputNodesFile != "none")
      {
        string outF = outputNodesFile  + "_" + TextTools::toString(itm.first);
        ApplicationTools::displayResult(" Output file for nodes", outF);

        ofstream out(outF.c_str(), ios::out);
        // map<int, vector<double> > frequencies;

        auto tree = pDR->getSubstitutionProcess()->getParametrizablePhyloTree();

        auto allIndex = addNodesExtant ? tree->getAllNodesIndexes() : tree->getAllInnerNodesIndexes();

        // Former output of bppAncestor


        vector<string> colNames;
        colNames.push_back("Nodes");
        // for (size_t i = 0; i < nbStates; i++)
        //   colNames.push_back("exp" + (sPP?sPP->getData()->getAlphabet()->intToChar((int)i):oPSP->getData()->getAlphabet()->intToChar((int)i)));
        for (size_t i = 0; i < nbStates; i++)
        {
          colNames.push_back("eb" + stateMap.getStateDescription(i));
        }

        // Now fill the table:
        vector<string> row(colNames.size());
        auto infos = make_unique<DataTable>(colNames);

        for (const auto& index:allIndex)
        {
          row[0] = TextTools::toString(index);

          Vdouble ebFreqs = sPP ? sPP->getPosteriorStateFrequencies(index) : oPSP->getPosteriorStateFrequencies(index);

          // for (size_t i = 0; i < nbStates; i++)
          // {
          //   row[i + 1] = TextTools::toString(itf.second[i]);
          // }
          for (size_t i = 0; i < nbStates; i++)
          {
            row[i + 1] = TextTools::toString(ebFreqs[i]);
          }
          infos->addRow(row);
        }

        DataTable::write(*infos, out, "\t");
      }

      ////////////////////////////////////////////////
      /// Ancestral sequences

      if (sequenceFilePath != "none")
      {
        shared_ptr<AlignedSequenceContainer> asSites = nullptr;

        // Write output:
        BppOAlignmentWriterFormat bppoWriter(1);
        unique_ptr<OAlignment> oAln(bppoWriter.read(sequenceFormat));
        ApplicationTools::displayResult(" Output alignment file ", sequenceFilePath + "_" + TextTools::toString(itm.first));
        ApplicationTools::displayResult(" Output alignment format ", oAln->getFormatName());

        if (sample)
        {
          asSites = make_shared<AlignedSequenceContainer>(alphabet);

          for (unsigned int i = 0; i < nbSamples; i++)
          {
            ApplicationTools::displayGauge(i, nbSamples - 1, '=');
            shared_ptr<AlignmentDataInterface> sampleSites = dynamic_cast<MarginalAncestralReconstruction*>(asr)->getAncestralSequences(true);
            vector<string> names = sampleSites->getSequenceNames();

            for (unsigned int j = 0; j < names.size(); j++)
            {
              names[j] += "_" + TextTools::toString(i + 1);
            }

            sampleSites->setSequenceNames(names, true);

            SequenceContainerTools::append(*asSites, *dynamic_pointer_cast<const SequenceContainerInterface>(sampleSites));
          }
          ApplicationTools::message->endLine();
        }
        else
          asSites = asr->getAncestralSequences();

        // Add existing sequence to output?
        if (addSitesExtant)
          SequenceContainerTools::append(*asSites, vSC);

        // Write sequences:
        oAln->writeAlignment(sequenceFilePath + "_" + TextTools::toString(itm.first), *asSites, true);
      }

      /// end of ancestral reconstruction
    }

    ApplicationTools::displayMessage("");
    bppancestor.done();
  }
  catch (exception& e)
  {
    cout << e.what() << endl;
    return 1;
  }

  return 0;
}
